include/deal.II/matrix_free/shape_info.h

```diff
@@ -275,6 +238,88 @@ namespace internal
        */
       Quadrature<1> quadrature;
 
+      /**
+       * Stores the degree of the element.
+       */
+      unsigned int fe_degree = 0;
+
+      /**
+       * Stores the number of quadrature points per dimension.
+       */
+      unsigned int n_q_points_1d = 0;
+
+      /**
+       * Indicates whether the basis functions are nodal in 0 and 1, i.e., the
+       * end points of the unit cell.
+       */
+      bool nodal_at_cell_boundaries = false;
+    };
+
+
+
+    /**
+     * The class that stores the shape functions, gradients and Hessians
+     * evaluated for a tensor product finite element and tensor product
+     * quadrature formula on the unit cell. Because of this structure, only
+     * one-dimensional data is stored.
+     *
+     * @ingroup matrixfree
+     *
+     * @author Katharina Kormann and Martin Kronbichler, 2010, 2011
+     */
+    template <typename Number>
+    struct ShapeInfo
+    {
+      /**
+       * Encodes the type of element detected at construction. FEEvaluation
+       * will select the most efficient algorithm based on the given element
+       * type.
+       */
+      ElementType element_type;
+
+      /**
+       * Empty constructor. Does nothing.
+       */
+      ShapeInfo();
+
+      /**
+       * Constructor that initializes the data fields using the reinit method.
+       */
+      template <int dim>
+      ShapeInfo(const Quadrature<1> &     quad,
+                const FiniteElement<dim> &fe,
+                const unsigned int        base_element = 0);
+
+      /**
+       * Initializes the data fields. Takes a one-dimensional quadrature
+       * formula and a finite element as arguments and evaluates the shape
+       * functions, gradients and Hessians on the one-dimensional unit cell.
+       * This function assumes that the finite element is derived from a one-
+       * dimensional element by a tensor product and that the zeroth shape
+       * function in zero evaluates to one.
+       */
+      template <int dim>
+      void
+      reinit(const Quadrature<1> &     quad,
+             const FiniteElement<dim> &fe_dim,
+             const unsigned int        base_element = 0);
+
+      /**
+       * Return data of univariate shape functions which defines the
+       * dimension @p dimension of tensor product shape functions
+       * regarding vector component @p component of the underlying
+       * finite element.
+       */
+      const UnivariateShapeData<Number> &
+      get_shape_data(const unsigned int dimension = 0,
+                     const unsigned int component = 0) const;
```

Comment:
This is something we still need to figure out how to do best - the access function adds another indirection with at least 2 instructions (load + chase pointer) for functions which are in the hot path. It will definitely not hurt higher-order cases, but I am not so sure about the 2d case with moderate degree without benchmarking things.